### ISSUE TITLE
[datagrid] Paging is not working when filter is applied

### ISSUE THREAD

**Issue by sivaah** [vmware/clarity#4752]
## Describe the bug

When filtering a column and clicking on next page shows the grid blank, It became blank even after removing the filters.

## How to reproduce

https://clarity-v3-datagrid-filter-paging.stackblitz.io/

![image](https://user-images.githubusercontent.com/6573919/85840551-47863800-b7ba-11ea-9c35-2ca9982e5f00.png)

1.  Go to https://clarity-v3-datagrid-filter-paging.stackblitz.io/
2.  Click on "Active" column and set filter to "true" or "false"
3.  Now click on next page
4.  see the grid is blank

Removing the filter also doesn't solve the problem.

## Expected behavior

It suppose to show the data for next page

## Versions

**App**

- Angular: 9.0.4
- Clarity: 3.1.4

**Device:**

- Type:MacBook
- OS: iOS
- Browser chrome
- Version 83.0.4103.116

## Additional notes

_Add any other notes about the problem here._

**Comment by kodemasta** [comment]
Tried sorting different columns by clicking on header and rows dissapear in stackblitz here !

**Comment by hippee-lee** [comment]
Thanks for reporting this. I think I've narrowed it down to something with the `[clrDgPageSize]` input. 
We will need to investigate more but this seems like a regression. 

What I saw is that when I start with 1 or 2 items / page its fine. Initializing with page size > 2 or any change to `clrDgPageSize` (selecting 3,4,5 etc) will break pagination.

I cleaned up the original example a bit while I was triaging. It still reproduced the error: 
https://stackblitz.com/edit/gh-4752-pagination-issue

I'm going to label this as a regression.

**Comment by kodemasta** [comment]
The issue is worse than described as simply sorting a column and paging and all rows are lost and unrecoverable short of a page reload. Need fix ASAP !

**Comment by hippee-lee** [comment]
I understand its urgent for you and we are aware of the issue. We will address it as soon as we can. We are a small distributed team so it will still take us a few days it investigate, fix and publish a patch release. 

As a workaround you can revert to v3.1.2. The issue was likely introduced by either https://github.com/vmware/clarity/pull/4682/ or https://github.com/vmware/clarity/pull/4626 

I suspect #4626.

**Comment by kodemasta** [comment]
I found a possible workaround for now...

Use a server driven callback to reload the row data. 
<clr-datagrid (clrDgRefresh)="refresh($event)" [clrDgLoading]="loading">

Thanks for looking into this !

**Comment by GaryHyland** [comment]
Hmm this is also a BIG issue for me in Clarity v2.4.4, Angular v8.3.25.
If have a datagrid with pagination and more rows than show on a page. sorting or filtering will work ok on the page showing, but breaks any pagination from then on. eg. same as the stackblitz shown. sort byb any column, then try and page forward. Broken.
Thankfully I have not deployed to PROD yet and will wait to see feedback.
Thanks..

**Comment by hippee-lee** [comment]
Thanks for the info. I'm marking this for v3 and v2 since there were several datagrid changes included in the releases since 2.4.2 & 3.1.2 more investigation is needed to determine which change(es) is the regression.

**Comment by Sanjeevraghav** [comment]
I faced the same issue when a Datagrid with pagination and more rows than shown on a page. If I clicked on any column for sort two times then all rows disappeared.
I faced this issue after upgrade Clarity v2.3.5 to v2.4.4 using ng update @clr/angular@2.4.4. & it also upgraded Angular v8.0.1 to v8.2.14.
But what I noticed this issue depends Angular and Clarity Version. 
1. Using ng update @clr/angular@version
    Angular v8.2.14 & Clarity v2.3.9  => Issue found
    Angular v8.2.14 & Clarity v2.4.2  => Issue found
    Angular v8.2.14 & Clarity v2.4.4  => Issue found

2. then I just upgrade Clarity using npm 
   Angular v8.0.1 & Clarity v2.3.5  => Issue found
   Angular v8.0.1 & Clarity v2.3.9  => No issue found
   Angular v8.0.1 & Clarity v2.4.2  => No issue found
   Angular v8.0.1 & Clarity v2.4.4  => Issue found

Note - I am using  Angular v8.0.1 & Clarity v2.4.2  to fix this issue.

**Comment by sivaah** [comment]
> I found a possible workaround for now...
> 
> Use a server driven callback to reload the row data.
> <clr-datagrid (clrDgRefresh)="refresh($event)" [clrDgLoading]="isLoading">
> 
> public refresh(state: ClrDatagridStateInterface) {
> this.isLoading = true;
> this.rowData = JSON.parse(JSON.stringify(this.rowData));
> this.isLoading = false;
> }
> 
> Thanks for looking into this !

The workaround brings another issue when we have clrDgSingleSelected, Whenever we navigate to new page or filter the selected is getting reset.

**Comment by kodemasta** [comment]
I do not see a fix in latest release 3.1.5. Any ETA for this ?  

For workaround we do deep clone of bound data during polling or manual refresh of grid. We also have to reassign selected grid row state as that is otherwise lost on paging forth then back.

Thanks

**Comment by Smorgasbordq** [comment]
I noticed that by detaching the displayed rows before inserting it, during datagrid's row.changes.subscription clears the issue.
I'm not sure why there are any displayedRows attached, though (and I'm guessing it used to be safe to assume the displayedRows were empty). It's got to be an issue during filtering.

```
this._subscriptions.push(
      this.rows.changes.subscribe(() => {
        if (!this.items.smart) {
          this.items.all = this.rows.map((row: ClrDatagridRow<T>) => row.item);
        }
        +for (let i = dg._displayedRows.length; i > 0; i--) {
        +  dg._displayedRows.detach();
        +}
        this.rows.forEach(row => {
          this._displayedRows.insert(row._view);
        });
      })
    );
```

**Comment by bbogdanov** [comment]
I was not able to reproduce this issue in v2. [Example](https://clarity-light-theme-v2-johgqt.stackblitz.io)

Version v3.1.5 has a fix this issue. I would encourage you to update to the next minor version.

**Comment by kodemasta** [comment]
Looks like 3.1.5 fixed issue. However, issue is back in 3.1.6. Regression. Changing page size large then small (10 to 20 then 10) then paging all data is lost in grid display.
